core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -434,6 +606,14 @@ public void updateTask(TaskResult taskResult) {
 
         String workflowId = taskResult.getWorkflowInstanceId();
         Workflow workflowInstance = executionDAO.getWorkflow(workflowId);
+
+
+        // FIXME Backwards compatibility for legacy workflows already running.
+        // This code will be removed in a future version.
+        if(workflowInstance.getWorkflowDefinition() == null) {
```

Comment:
This logic should be encapsulated in ExecutionDAO which automatically populates workflowDefinition and taskDefinitions for legacy workflows, now this is repeated in several places.

Reply:
ExecutionDAO does not populate legacy workflows anymore, this was done intentionally. This class is only in charge of running and creating objects related to workflow and task executions.
 
If the problem is that you would like to see the population of the workflowDefinition on the ExecutionDAO layer, the code on that layer will just be a proxy to the MetadataMapperService.

However I would not recommend that because:
- Lets say we have a proxy method on the executionDao layer. We are going to be adding a deprecated method that consists on being a proxy to another method. On top of that we are likely to have duplication of code on each of the implementations of executionDAO.
- The way we start a workflow is loading the metadata on WorkflowExecutor and running decisions on it. That was the main idea behind immutable workflow definitions during workflow execution.
- When we need to update a legacy task, we load the metadata and run the decisions on it. At the moment of running decisions, the task definition for that task is immutable. Achieving similar benefits as a new workflow execution.

Having said that, if it does not convince the idea of use MetadataMapperService, we can create a new method for legacy reasons on the ExecutionDAO layer and proxy to metadataMapperService.